### PR TITLE
Elixir Provider

### PR DESCRIPTION
**Pull request by zoedsoupe**
I'm trying to test some possibilities but I also think that's important to other people review this code.

It's on very early stage!

### CODE REVIEW

**Comment by JakeCooper** [comment]
Hey aweseome! Let us know once it's fully ready for review and one of us can have a look :)

Thanks so much (in advance) for contributing :D

**Comment by nebulatgs** [comment]
Fixes #75

**Comment by zoedsoupe** [comment]
> Hey awesome! Let us know once it's fully ready for review and one of us can have a look :)
> 
> Thanks so much (in advance) for contributing :D

Indeed! Need to implement nervers build branch and start phase and then some refactoring to follow Heroku's buildpack :)

This is my first OSS contribution and I'm really liking

File: .gitignore [hunk 1]
```diff
@@ -53,3 +52,6 @@ tmp/
 testing
 nix-testing
 test
+
+# Ignore nix build
+result
```
Comment (Milo123459) [inline .gitignore:57]: All changes in this file look to be unrelated and possibly related to your other pr #180

File: flake.lock [hunk 2]
```diff
@@ -0,0 +1,61 @@
+{
```
Comment (Milo123459) [inline flake.lock:1]: This file also seems to be related to your other PR.

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
+{
```
Comment (Milo123459) [inline flake.nix:1]: .... and this.

File: src/providers/elixir.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+pub struct ElixirProvider {}
+
+const AVAILABLE_ELIXIR_VERSIONS: &[(f64, &str)] = &[
+    (1.9, "elixir_1_9"),
+    (1.10, "elixir_1_10"),
+    (1.11, "elxir_1_11"),
+    (1.12, "elixir_1_12"),
+    (1.13, "elixir"),
+];
+const ELIXIR_VERSION_TO_OTP_PKG: &[(f64, &str)] = &[
```
Comment (Milo123459) [inline src/providers/elixir.rs:39]:
```suggestion
const ELIXIR_VERSION_TO_OTP_PKG: &[(f32, &str)] = &[
```

File: src/providers/elixir.rs [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        let erlang_nix_pkg = ElixirProvider::get_nix_erlang_pkg(mix_project)?;
+
+        match mix_project {
+            Some(MixProject {
+                is_escript: true, ..
+            }) => {
+                let pkgs = vec![Pkg::new(&elixir_nix_pkg), Pkg::new(&erlang_nix_pkg)];
+
+                return Ok(Some(SetupPhase::new(pkgs)));
+            }
```
Comment (Milo123459) [inline src/providers/elixir.rs:72]:
```suggestion
            }) => Ok(Some(SetupPhase::new(vec![Pkg::new(&elixir_nix_pkg), Pkg::new(&erlang_nix_pkg)]))),
```

File: src/providers/elixir.rs [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+            _project => Ok(Some(SetupPhase::new(vec![Pkg::new(&elixir_nix_pkg)]))),
```
Comment (Milo123459) [inline src/providers/elixir.rs:74]:
```suggestion
            _ => Ok(Some(SetupPhase::new(vec![Pkg::new(&elixir_nix_pkg)]))),
```

**Review by Milo123459 — CHANGES_REQUESTED** [review]
You need to add your tests and remove the unrelated changes from your other PR. 

Other than that, so far so good!

**Comment by Milo123459** [comment]
> > You need to add your tests and remove the unrelated changes from your other PR.
> > Other than that, so far so good!
> 
> I think the best course of action would be to rebase this branch on `main` which doesn't contain the extra commit.

It's fine to have an extra commit..

File: flake.lock (hunk 2 above)
Comment (coffee-cup) [reply flake.lock:1]: +1 This shouldn't be added to this PR

File: src/providers/elixir.rs [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+        let closest_version = AVAILABLE_ELIXIR_VERSIONS.iter().find(|(v, _)| v >= version);
+
+        if let Some((_, pkg)) = closest_version {
+            return Ok(Some(pkg.to_string()));
+        }
+
+        Ok(None)
+    }
+
+    fn parse_mix_project(mix_exs: Option<String>) -> Result<Option<MixProject>> {
```
Comment (coffee-cup) [inline src/providers/elixir.rs:249]: Suggestion: Add a test or two for this parser

**Review by coffee-cup — COMMENTED** [review]
This seems to be coming along nicely! I think there should also be an `elixir-phoenix` example that we can test against too. With some automated tests for both vanilla and phoenix

**Comment by JakeCooper** [comment]
Hey @zoedsoupe any progress here?

**Comment by zoedsoupe** [comment]
Yeah! I got stuck in these recents days because of my job, but I'll update this branch asap.

**Comment by JakeCooper** [comment]
@zoedsoupe Closing to cleanup PRs. Feel free to re-open once ready
